// @temporalio/worker/lib/workflow-log-interceptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interceptors = exports.WorkflowInboundLogInterceptor = exports.workflowLogAttributes = void 0;
const workflow_1 = require("@temporalio/workflow");
const stack_helpers_1 = require("@temporalio/workflow/lib/stack-helpers");
/**
 * Returns a map of attributes to be set on log messages for a given Workflow
 */
function workflowLogAttributes(info) {
    return {
        namespace: info.namespace,
        taskQueue: info.taskQueue,
        workflowId: info.workflowId,
        runId: info.runId,
        workflowType: info.workflowType,
    };
}
exports.workflowLogAttributes = workflowLogAttributes;
/** Logs Workflow execution starts and completions */
class WorkflowInboundLogInterceptor {
    logAttributes() {
        return workflowLogAttributes((0, workflow_1.workflowInfo)());
    }
    execute(input, next) {
        const { defaultWorkerLogger: logger } = (0, workflow_1.proxySinks)();
        logger.debug('Workflow started', this.logAttributes());
        const p = next(input).then((res) => {
            logger.debug('Workflow completed', this.logAttributes());
            return res;
        }, (error) => {
            // Avoid using instanceof checks in case the modules they're defined in loaded more than once,
            // e.g. by jest or when multiple versions are installed.
            if (typeof error === 'object' && error != null) {
                if ((0, workflow_1.isCancellation)(error)) {
                    logger.debug('Workflow completed as cancelled', this.logAttributes());
                    throw error;
                }
                else if (error.name === 'ContinueAsNew') {
                    logger.debug('Workflow continued as new', this.logAttributes());
                    throw error;
                }
            }
            logger.warn('Workflow failed', { error, ...this.logAttributes() });
            throw error;
        });
        // Avoid showing this interceptor in stack trace query
        (0, stack_helpers_1.untrackPromise)(p);
        return p;
    }
}
exports.WorkflowInboundLogInterceptor = WorkflowInboundLogInterceptor;
// ts-prune-ignore-next
const interceptors = () => ({ inbound: [new WorkflowInboundLogInterceptor()] });
exports.interceptors = interceptors;
