// @temporalio/common/lib/type-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorCode = exports.errorMessage = exports.hasOwnProperties = exports.hasOwnProperty = exports.isRecord = exports.checkExtends = void 0;
/** Verify that an type _Copy extends _Orig */
function checkExtends() {
    // noop, just type check
}
exports.checkExtends = checkExtends;
function isRecord(value) {
    return typeof value === 'object' && value !== null;
}
exports.isRecord = isRecord;
// ts-prune-ignore-next
function hasOwnProperty(record, prop) {
    return prop in record;
}
exports.hasOwnProperty = hasOwnProperty;
function hasOwnProperties(record, props) {
    return props.every((prop) => prop in record);
}
exports.hasOwnProperties = hasOwnProperties;
/**
 * Get `error.message` (or `undefined` if not present)
 */
function errorMessage(error) {
    if (typeof error === 'string') {
        return error;
    }
    if (error instanceof Error) {
        return error.message;
    }
    return undefined;
}
exports.errorMessage = errorMessage;
/**
 * Get `error.code` (or `undefined` if not present)
 */
function errorCode(error) {
    if (typeof error === 'object' &&
        error.code !== undefined &&
        typeof error.code === 'string') {
        return error.code;
    }
    return undefined;
}
exports.errorCode = errorCode;
